### ISSUE TITLE
Disassembler outputs addresses in decimal

### ISSUE THREAD

**Issue by pfusik** [llvm-mos/llvm-mos#242]
Environment:

    C:\0\a8\llvm>install\bin\llvm-objdump --version
    LLVM (http://llvm.org/):
      LLVM version 16.0.0git
      Optimized build.
      Default target: mos
      Host CPU: znver1

      Registered Targets:
        mos - MOS Technologies 65xx and variants

Steps to reproduce:

    C:\0\a8\llvm>cat main.c
    int main() {
    }

    C:\0\a8\llvm>install\bin\mos-atari8-clang -Os -Wall -o main.xex main.c

    C:\0\a8\llvm>install\bin\llvm-objdump -d main.xex.elf

    main.xex.elf:   file format elf32-mos

    Disassembly of section .text:

    00002000 <_start>:
        2000: 20 11 20      jsr     8209

    00002003 <__after_main>:
        2003: 4c 07 20      jmp     8199

    00002006 <_fini>:
        2006: 60            rts

    00002007 <__exit_return>:
        2007: 8d 15 20      sta     8213
        200a: 20 06 20      jsr     8198
        200d: ad 15 20      lda     8213
        2010: 60            rts

    00002011 <main>:
        2011: a2 00         ldx     #0
        2013: 8a            txa
        2014: 60            rts

    C:\0\a8\llvm>install\bin\llvm-objdump -d --print-imm-hex main.xex.elf

    main.xex.elf:   file format elf32-mos

    Disassembly of section .text:

    00002000 <_start>:
        2000: 20 11 20      jsr     $2011

    00002003 <__after_main>:
        2003: 4c 07 20      jmp     $2007

    00002006 <_fini>:
        2006: 60            rts

    00002007 <__exit_return>:
        2007: 8d 15 20      sta     $2015
        200a: 20 06 20      jsr     $2006
        200d: ad 15 20      lda     $2015
        2010: 60            rts

    00002011 <main>:
        2011: a2 00         ldx     #$0
        2013: 8a            txa
        2014: 60            rts

Expected:

Addresses shown in hexadecimal by default, as in all 6502 disassemblers ever.
`--print-imm-hex` only affects the immediate addressing mode.

Actual:

All disassembly operands shown in decimal by default.
This makes the listing barely human-readable. For instance, one must understand that 8209 == 0x2011.
`--print-imm-hex` is a workaround.

**Comment by mysterymath** [comment]
The current semantics of the `--print-imm-hex` flag agrees with the behavior of llvm-objdump for e.g. x86:
```
[protoman ~]$ cat tmp.c
void foo(void) {
  volatile char *x = (volatile char *)0x1234;
  *x = 42;
}
[protoman ~]$ clang -O2 -c tmp.c
[protoman ~]$ llvm-objdump -d tmp.o

tmp.o:	file format elf64-x86-64

Disassembly of section .text:

0000000000000000 <foo>:
       0: c6 04 25 34 12 00 00 2a      	movb	$42, 4660
       8: c3                           	retq

[protoman ~]$ llvm-objdump -d --print-imm-hex tmp.o

tmp.o:	file format elf64-x86-64

Disassembly of section .text:

0000000000000000 <foo>:
       0: c6 04 25 34 12 00 00 2a      	movb	$0x2a, 0x1234
       8: c3                           	retq
```

Still, we have the latitude to make --print-imm-hex the default behavior for MOS without damaging the expected semantics of llvm-objdump, and as you've mentioned, this is the overwhelming default behavior for 6502 assemblers.

**Comment by mysterymath** [comment]
You've uncovered quite a comedy of errors here; it looks like the objdump utility that llvm-objdump is emulating defaults to hex for all platforms, and there was a change in 2016 to make LLVM this way that got through code review and was submitted: https://reviews.llvm.org/D18770

However, they forgot to fix some LLD tests that broke due to this, so they rolled it back: https://github.com/llvm/llvm-project/commit/efe37328831b7184548e038f067e3dd8313ec42f

It looks like nobody every actually went back and fixed up all the tests to get this change back in. This seems like absolutely the best way to fix this, since hex is generally more useful for other targets as well. I'll try to push this through in upstream LLVM, then fetch the changes to resolve this.

**Comment by pfusik** [comment]
Oh. I keep my fingers crossed!

**Comment by mysterymath** [comment]
The upstream change is out for review: https://reviews.llvm.org/D136972.

**Comment by mysterymath** [comment]
The change has been LGTMed! I'll work on committing it tomorrow, with fresh eyes.

**Comment by mysterymath** [comment]
This has now landed upstream as https://github.com/llvm/llvm-project/commit/cc2457ca1bbdddefc09d72f889bc0389c90a831e.

I'm now starting work on an integration from upstream.

**Comment by mysterymath** [comment]
Fixed by cc2457ca
